File: pingpong-typescript-sdk/src/lib/deployments/deployments.ts
import { Client } from '../client/client.js';
import { Models } from '../models/models.js';
import { Deployment, DeploymentInput, Job } from './model.js';

const COMPLETE = 'complete';
const FAILED = 'failed';
const DOCKER_PULLING = "docker_pulling";
const STARTING = 'starting'

interface DeploymentInputWithIndex extends DeploymentInput {
    [key: string]: unknown;
}

class Deployments extends Client {
    private modelsClient: Models;

    constructor(apiKey: string, modelClient: Models) {
        super(apiKey);
        this.modelsClient = modelClient;
    }

    async create(deployment: DeploymentInputWithIndex): Promise<Deployment | undefined> {
        let model;
        try {
            model = await this.modelsClient.getById(deployment.model);
        } catch (e) {
            if (e instanceof Error) {
                throw new Error(`error fetching model: ${e.message}`);
            }
        }
        deployment.model_id = model?.id;
        try {
            let actualResponse = await this.post<DeploymentInputWithIndex, Deployment>(
                '/api/v1/deployments',
                deployment,
            );
            if (actualResponse != undefined) {
                let eta = (actualResponse.job.eta * 3) + 120;
                if (deployment.sync) {
                    while (actualResponse.status !== COMPLETE && actualResponse.status !== FAILED && eta > 0) {
                        await this.sleep(10 * 1000);
                        let job = await this.getJob(actualResponse.job_id)
                        if (job != undefined) {
                            if (job.status !== DOCKER_PULLING && job.status != STARTING) {
                                eta -= 10;
                            }
                            actualResponse.job = job;
                            actualResponse.logs = job.logs;
                            actualResponse.status = job.status;
                        }
                    }
                }
            }
            return actualResponse;
        } catch (e) {
            if (e instanceof Error) {
                throw new Error(`error creating deployment: ${e.message}`);
            }
        }
    }

    async list(): Promise<Deployment[] | undefined> {
        try {
            return this.get<Deployment[]>("/api/v1/deployments");
        } catch (e: unknown) {
            if (e instanceof Error) {
                throw new Error(`error listing deployments: ${e.message}`);
            }
        }
    }

    async getJob(id: string): Promise<Job | undefined> {
        try {
            return this.get<Job>(`/api/v1/jobs/${id}`);
        } catch (e: unknown) {
            if (e instanceof Error) {
                throw new Error(`error getting job status: ${e.message}`);
            }
        }
    }
    async sleep(ms: number): Promise<void> {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

export { Deployments };

File: pingpong-python-sdk/pingpongsdk/deployments/deployments.py
from ..shared.client.client import Client
from .model import Deployment, dto, request_dto, CreateDeployment,Job
import time
from typing import List

from dataclasses import asdict


class Deployments(Client):

    def __init__(self, api_key: str, models) -> None:
        self.models = models
        super().__init__(api_key)


    def get_by_id(self, id: str) -> Deployment:
        """
        get_by_id - get a deployment by its ID
        """
        deployment = super().get("/api/v1/deployments/%s" % id)
        return dto(deployment)
    

    def create(self, deployment: CreateDeployment) -> Deployment:
        """
        create - create a new Deployment
        """
        try:
            model = self.models.get_by_id(deployment.model)
            id = model.id
            request = request_dto(id, deployment)
            sync=deployment.sync
            deployment = super().post("/api/v1/deployments", request)
            response=dto(deployment)
            status=response.status
            eta= response.job['eta']
            if sync: 
                while status!='complete' and status!='failed' and eta > 0:
                    time.sleep(10)
                    job=self.get_job(response.job_id)
                    eta-=5
                    status=job['status']
                    response.status=status
                    response.logs=job['logs']
                    response.job=job
            return response
        except Exception as e:
            raise Exception('error creating deployment %s' % e)


    def list(self, start: int, to: int) -> List[Deployment]:
        """
        list - list all of your deployments
        """
        deployments = super().get("/api/v1/deployments", start, to)

        result = []
        for d in deployments:
            result.append(dto(d))

        return result
    
    def get_job(self, id:str) -> Job:
        """
        get_job - get job by ID
        """
        path="/api/v1/jobs/"+ id
        job = super().get(path)

        return job

    
